demo/runtime/src/dispatch.rs

```diff
@@ -16,79 +16,230 @@
+								$(
+									ref $param_name
+								),*
+							) => {
+								(Id::$fn_name as u8).using_encoded(|s| v.extend(s));
+								$(
+									$param_name.using_encoded(|s| v.extend(s));
+								)*
+							}
+						)*
+					}
+					v
+				}
+
+				fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+					f(self.encode().as_slice())
+				}
+			}
+			impl $crate::dispatch::NonTrivialSlicable for Call {}
+		}
 	}
 }
 
-/// Dispatch a function.
-pub fn function(function: &Function, transactor: &AccountId) {
-	match *function {
-		Function::StakingStake =>
-			staking::public::stake(transactor),
-		Function::StakingUnstake =>
-			staking::public::unstake(transactor),
-		Function::StakingTransfer(dest, value) =>
-			staking::public::transfer(transactor, &dest, value),
-		Function::SessionSetKey(session) =>
-			session::public::set_key(transactor, &session),
-		Function::TimestampSet(t) =>
-			timestamp::public::set(t),
-		Function::CouncilVotePropose(ref a) =>
-			council_vote::public::propose(transactor, a),
-		Function::CouncilVoteVote(ref a, b) =>
-			council_vote::public::vote(transactor, a, b),
-		Function::CouncilVoteVeto(ref a) =>
-			council_vote::public::veto(transactor, a),
-		Function::CouncilSetApprovals(ref a, b) =>
-			council::public::set_approvals(transactor, a, b),
-		Function::CouncilReapInactiveVoter(a, ref b, c, d) =>
-			council::public::reap_inactive_voter(transactor, a, b, c, d),
-		Function::CouncilRetractVoter(a) =>
-			council::public::retract_voter(transactor, a),
-		Function::CouncilSubmitCandidacy(a) =>
-			council::public::submit_candidacy(transactor, a),
-		Function::CouncilPresentWinner(ref a, b, c) =>
-			council::public::present_winner(transactor, a, b, c),
-		Function::DemocracyPropose(ref a, b) =>
-			democracy::public::propose(transactor, a, b),
-		Function::DemocracySecond(a) =>
-			democracy::public::second(transactor, a),
-		Function::DemocracyVote(a, b) =>
-			democracy::public::vote(transactor, a, b),
+macro_rules! impl_meta_dispatch {
+	(
+		pub mod $super_name:ident;
+		path $path:ident;
+		trait $trait:ty;
+		$(
+			$camelcase:ident(mod $sub_name:ident) = $id:expr ;
+		)*
+	) => {
+		pub mod $super_name {
+			use super::*;
+
+			#[derive(Clone, Copy, PartialEq, Eq)]
+			#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+			#[repr(u32)]
+			#[allow(non_camel_case_types)]
+			enum Id {
+				$(
+					#[allow(non_camel_case_types)]
+					$camelcase = $id,
+				)*
+			}
+
+			impl Id {
+				/// Derive `Some` value from a `u8`, or `None` if it's invalid.
+				fn from_u8(value: u8) -> Option<Id> {
+					match value {
+						$(
+							$id => Some(Id::$camelcase),
+						)*
+						_ => None,
+					}
+				}
+			}
+
+			#[derive(Clone, PartialEq, Eq)]
+			#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+			#[allow(missing_docs)]
+			pub enum Call {
+				$(
+					#[allow(non_camel_case_types)]
+					$camelcase ( $crate::runtime::$sub_name::$path::Call )
```

Comment:
this is a little non-general. If we want this to be reusable for other runtimes we shouldn't make any assumptions about module structure.